pkg/datapath/loader/hash_test.go

```diff
@@ -4,101 +4,108 @@
-	h.Reset()
-	require.Equal(t, h.String(), baseHash)
-	require.NotEqual(t, h.String(), dummyHash)
-
-	// Ensure that with a copy of the endpoint config we get the same hash
-	newEPCfg := dummyEPCfg
-	h = hashDatapath(cfg, &dummyNodeCfg, &dummyDevCfg, &newEPCfg)
-	require.NotEqual(t, h.String(), baseHash)
-	require.Equal(t, h.String(), dummyHash)
+	// Error from ConfigWriter is forwarded.
+	_, err := hashDatapath(fakeConfigWriter{}, nil)
+	require.Error(t, err)
+
+	// Ensure we get different hashes when config is changed
+	a, err := hashDatapath(fakeConfigWriter("a"), &dummyNodeCfg)
+	require.NoError(t, err)
+
+	b, err := hashDatapath(fakeConfigWriter("b"), &dummyNodeCfg)
+	require.NoError(t, err)
+	require.NotEqual(t, a, b)
+
+	// Ensure we get the same base hash when config is the same.
+	b, err = hashDatapath(fakeConfigWriter("a"), &dummyNodeCfg)
+	require.NoError(t, err)
+	require.Equal(t, a, b)
+}
+
+func TestHashEndpoint(t *testing.T) {
+	var base datapathHash
+	ep := testutils.NewTestEndpoint()
+	cfg := configWriterForTest(t)
+
+	// Error from ConfigWriter is forwarded.
+	_, err := base.hashEndpoint(fakeConfigWriter{}, nil)
+	require.Error(t, err)
+
+	// Hashing the endpoint gives a hash distinct from the base.
+	a, err := base.hashEndpoint(cfg, &ep)
+	require.NoError(t, err)
+	require.NotEqual(t, base.String(), a)
+
+	// When we configure the endpoint differently, it's different
+	ep.Opts.SetBool("foo", true)
+	b, err := base.hashEndpoint(cfg, &ep)
+	require.NoError(t, err)
+	require.NotEqual(t, a, b)
+}
+
+func TestHashTemplate(t *testing.T) {
+	var base datapathHash
+	ep := testutils.NewTestEndpoint()
+	cfg := configWriterForTest(t)
+
+	// Error from ConfigWriter is forwarded.
+	_, err := base.hashTemplate(fakeConfigWriter{}, nil)
+	require.Error(t, err)
+
+	// Hashing the endpoint gives a hash distinct from the base.
+	a, err := base.hashTemplate(cfg, &ep)
+	require.NoError(t, err)
+	require.NotEqual(t, base.String(), a)
 
 	// Even with different endpoint IDs, we get the same hash
 	//
 	// This is the key to avoiding recompilation per endpoint; static
 	// data substitution is performed via pkg/elf instead.
-	newEPCfg.Id++
-	h = hashDatapath(cfg, &dummyNodeCfg, &dummyDevCfg, &newEPCfg)
-	require.NotEqual(t, h.String(), baseHash)
-	require.Equal(t, h.String(), dummyHash)
-
-	// But when we configure the endpoint differently, it's different
-	newEPCfg = testutils.NewTestEndpoint()
-	newEPCfg.Opts.SetBool("foo", true)
-	h = hashDatapath(cfg, &dummyNodeCfg, &dummyDevCfg, &newEPCfg)
-	require.NotEqual(t, h.String(), baseHash)
-	require.NotEqual(t, h.String(), dummyHash)
+	ep.Id++
+	b, err := base.hashTemplate(cfg, &ep)
+	require.NoError(t, err)
+	require.Equal(t, a, b)
+}
+
+type fakeConfigWriter []byte
+
+func (fc fakeConfigWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeConfiguration) error {
+	if cfg == nil {
+		return errors.New("LocalNodeConfiguration is nil")
+	}
+	_, err := w.Write(fc)
```

Comment:
This is necessary because it's hard to make the real ConfigWriter not return an error due to missing network devices. Tight coupling for the win.